src/Substrait.Core/ProtoConverter.cs

```diff
@@ -0,0 +1,154 @@
+  /// </summary>
+  public class ProtoConverter
+  {
+    /// <summary>
+    /// Create domain object specific to relational operator type from protobuf message
+    /// </summary>
+    /// <param name="protoRel">Relation encoded as protobuf message</param>
+    /// <returns><see cref="Rel"/> corresponding to message's operator type</returns>
+    public Rel ToRel(ProtoRel protoRel)
+    {
+      var relType = protoRel.RelTypeCase;
+
+      return relType switch
+      {
+        ProtoRel.RelTypeOneofCase.Aggregate => CreateAgg(protoRel.Aggregate),
+        ProtoRel.RelTypeOneofCase.Cross => CreateCross(protoRel.Cross),
+        ProtoRel.RelTypeOneofCase.Fetch => CreateFetch(protoRel.Fetch),
+        ProtoRel.RelTypeOneofCase.Filter => CreateFilter(protoRel.Filter),
+        ProtoRel.RelTypeOneofCase.Join => CreateJoin(protoRel.Join),
+        ProtoRel.RelTypeOneofCase.Project => CreateProject(protoRel.Project),
+        ProtoRel.RelTypeOneofCase.Read => CreateRead(protoRel.Read),
+        ProtoRel.RelTypeOneofCase.Sort => CreateSort(protoRel.Sort),
+        _ => throw new NotImplementedException(relType.ToString()),
+      };
+    }
+
+    /// <summary>
+    /// Create domain object specific to relational expression type from protobuf message
+    /// </summary>
+    /// <param name="protoRex">Relational expression encoded as protobuf message</param>
+    /// <returns><see cref="Rel"/> corresponding to message's expression type</returns>
+    private RelExpression ToExpression(ProtoRex protoRex)
+    {
+      var rexType = protoRex.RexTypeCase;
+
+      return rexType switch
+      {
+        ProtoRex.RexTypeOneofCase.Selection => CreateFieldReference(protoRex.Selection),
+        ProtoRex.RexTypeOneofCase.None => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Literal => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.ScalarFunction => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.WindowFunction => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.IfThen => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.SwitchExpression => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.SingularOrList => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.MultiOrList => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Cast => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Subquery => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Nested => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Enum => throw new NotImplementedException(),
+        _ => throw new NotImplementedException(rexType.ToString()),
+      };
+    }
+
+    private Rel CreateAgg(AggregateRel agg)
+    {
+      var input = ToRel(agg.Input);
+      var inputs = ImmutableList.Create(input);
```

Comment:
It seems there are important parts missing in the translation, e.g., grouping columns or aggregates. Same comment for other operators below (e.g., expressions for join, filter, or project, different type of read rels, etc.)